### PR TITLE
refactor webhookAnalytics call and enrich analytics module

### PR DESCRIPTION
**Pull request by anamarn**
**TLDR**

Refactor WebhoonAnalytics Graph to a more abstract version AnalyticsGraph (in analytics module). Thus enabling the components to be used on different instances (ex: new endpoint, new kind of graph).

**In order to test:**

1. Set ANALYTICS_ENABLED to true
2. Set TINYBIRD_JWT_TOKEN to the ADMIN token from the workspace twenty_analytics_playground
3. Set TINYBIRD_JWT_TOKEN to the datasource or your admin token from the workspace twenty_analytics_playground
4. Create a Webhook in twenty and set wich events it needs to track
5. Run twenty-worker in order to make the webhooks work.
6. Do your tasks in order to populate the data
7. Enter to settings> webhook>your webhook and the statistics section should be displayed.

### CODE REVIEW

**Comment by FelixMalfait** [comment]
Great, starting to review!

Todo: check why it's loaded 3 times on page load
Screenshot 2024-11-01 at 13 17 17

File: packages/twenty-ui/tsconfig.json [hunk 1]
```diff
@@ -17,16 +17,20 @@
   "include": [],
   "references": [
     {
-      "path": "./tsconfig.build.json"
+      "path": "./tsconfig.build.json",
+      "composite": true
```
Comment (FelixMalfait) [inline packages/twenty-ui/tsconfig.json:21]: why do this change? does it come from somewhere else?

File: packages/twenty-front/src/pages/settings/developers/webhooks/components/SettingsDevelopersWebhookDetail.tsx [hunk 2]
```diff
@@ -282,8 +282,16 @@ export const SettingsDevelopersWebhooksDetail = () => {
...
-            <SettingsDevelopersWebhookUsageGraphEffect webhookId={webhookId} />
-            <SettingsDevelopersWebhookUsageGraph webhookId={webhookId} />
+            <AnalyticsGraphEffect
+              recordId={webhookId}
+              recordType="webhook"
+              endpointName="getWebhookAnalytics"
+            />
+            <AnalyticsActivityGraph
+              recordId={webhookId}
+              recordType="webhook"
```
Comment (FelixMalfait) [inline packages/twenty-front/src/pages/settings/developers/webhooks/components/SettingsDevelopersWebhookDetail.tsx:292]: We can have a map (constant) that stores the record type for each endpoint, instead of passing it downstream here. GetWebhookAnalytics will always be with the webhook recordType.

File: packages/twenty-front/src/modules/analytics/utils/fetchGraphDataOrThrow.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  windowLength,
+  tinybirdJwt,
+  endpointName,
+}: fetchGraphDataOrThrowProps) => {
+  const queryString = new URLSearchParams({
+    ...ANALYTICS_GRAPH_OPTION_MAP[windowLength],
+    ...computeStartEndDate(windowLength),
+    ...(recordType === 'webhook'
+      ? { webhookId: recordId }
+      : { functionId: recordId }), //later create a switch case
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/analytics/utils/fetchGraphDataOrThrow.ts:23]: concatenate recordType + Id (recordType which you should get from a constant map, cf other comment)

File: packages/twenty-front/src/modules/analytics/constants/AnalyticsGraphOptionMapCustom.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+//Use this map when granularity is custom meanwhile i'll use the first map
+export const ANALYTICS_GRAPH_OPTION_MAP_CUSTOM = {
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/analytics/constants/AnalyticsGraphOptionMapCustom.ts:2]: Do we need this? Or can we remove it before merging?

File: packages/twenty-front/src/modules/analytics/hooks/useAnalyticsGraphDataState.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { webhookAnalyticsGraphDataState } from '@/analytics/states/webhookAnalyticsGraphDataState';
+
+import { mapWebhookAnalyticsResultToNivoLineInput } from '@/analytics/utils/mapWebhookAnalyticsResultToNivoLineInput';
+
+export const useAnalyticsGraphDataState = (endpointName: string) => {
+  switch (endpointName) {
+    case 'getWebhookAnalytics':
+      return {
+        analyticsState: webhookAnalyticsGraphDataState,
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/analytics/hooks/useAnalyticsGraphDataState.ts:9]: It feels like what you want to use here might be componentState? Look for `createComponentState` in the code :)

File: packages/twenty-front/src/pages/settings/serverless-functions/SettingsServerlessFunctionDetail.tsx [hunk 6]
```diff
@@ -169,11 +178,19 @@ export const SettingsServerlessFunctionDetail = () => {
...
+
+  const isAnalyticsV2Enabled = useIsFeatureEnabled('IS_ANALYTICS_V2_ENABLED');
 
   const tabs = [
     { id: 'editor', title: 'Editor', Icon: IconCode },
     { id: 'test', title: 'Test', Icon: IconTestPipe },
     { id: 'settings', title: 'Settings', Icon: IconSettings },
+    ...(isAnalyticsEnabled && isAnalyticsV2Enabled
+      ? [
+          { id: 'monitoring', title: 'Monitoring', Icon: IconBox }, //IconGauge from figma is not in twenty ui:(
```
Comment (FelixMalfait) [inline packages/twenty-front/src/pages/settings/serverless-functions/SettingsServerlessFunctionDetail.tsx:191]: You can add any tabler icon to twenty-ui :). I think you just need to add it to `TablerIcons.ts` in twenty-ui

File: packages/twenty-front/src/modules/analytics/components/AnalyticsActivityGraph.tsx [hunk 7]
```diff
@@ -33,34 +29,39 @@ const StyledTitleContainer = styled.div`
...
-  const { fetchGraphData } = useGraphData(webhookId);
+  const { fetchGraphData } = useGraphData({
+    recordId,
+    endpointName,
+  });
+
+  const transformDataFunction = computeAnalyticsGraphDataFunction(endpointName);
 
+  const dropdownId = useId();
+  // perhaps here i need to separate the Section container and the graph itself? TODO: Add elements of distintion btwen graphs of the same record type
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/analytics/components/AnalyticsActivityGraph.tsx:53]: Could you remove this comment?

File: packages/twenty-front/src/modules/analytics/components/AnalyticsActivityGraph.tsx [hunk 8]
```diff
@@ -33,34 +29,39 @@ const StyledTitleContainer = styled.div`
...
     <>
-      {webhookGraphData.length ? (
+      {analyticsGraphData.length ? (
         <Section>
           <StyledTitleContainer>
             <H2Title
-              title="Activity"
-              description="See your webhook activity over time"
+              title={`${ANALYTICS_GRAPH_TITLE_MAP[endpointName]}`}
+              description={`See your ${ANALYTICS_ENDPOINT_TYPE_MAP[endpointName]} activity over time`}
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/analytics/components/AnalyticsActivityGraph.tsx:61]: Let's stay consistent and use a map for descriptions, just like you did for title. This will give us more flexibility to customize the text. (note we could have also passed title/description as parameter of the component, that's how I would have done it intuitively but the way you did it works too!)
